First, import utilities.

In [1]:
import tensorflow as tf

Load the MNIST dataset.

In [2]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

11493376/11490434 [==============================] - 39s 3us/step


Set up the network with Keras. The Sequential method takes array of the layers of the neural network. A model is a concrete neural network. A neural network is comprised of different layer types.

In [5]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)
])

Get the prediction scores for each class.

In [6]:
predictions = model(x_train[:1]).numpy()
predictions


If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.



array([[ 0.26522306, -0.07683796, -0.04950114, -0.13446873,  0.38292986,
         0.05236054, -1.1303933 ,  0.1698519 ,  0.6241049 , -0.19714272]],
      dtype=float32)

These scores can now be converted to probabilities.

In [7]:
tf.nn.softmax(predictions).numpy()

array([[0.12103259, 0.08597003, 0.0883526 , 0.08115558, 0.13615127,
        0.09782667, 0.02997739, 0.11002292, 0.1732857 , 0.07622534]],
      dtype=float32)

To get the losses, the SparseCategoricalCrossentropy method is called with a flag true on the parameter from_logits which are the prediction scores.
The loss is the negative score of the probability.

In [8]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

This neural network is untrained.

In [9]:
loss_fn(y_train[:1], predictions).numpy()

2.324558

Configure the model with optimzers, metrics and initial losses.

In [11]:
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

Train the model with the fit method.

In [12]:
model.fit(x_train, y_train, epochs=5)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 10s 174us/sample - loss: 0.2960 - accuracy: 0.9138
Epoch 2/5
60000/60000 [==============================] - 10s 169us/sample - loss: 0.1409 - accuracy: 0.9574
Epoch 3/5
60000/60000 [==============================] - 10s 171us/sample - loss: 0.1070 - accuracy: 0.9672
Epoch 4/5
60000/60000 [==============================] - 12s 195us/sample - loss: 0.0868 - accuracy: 0.9727
Epoch 5/5
60000/60000 [==============================] - 12s 196us/sample - loss: 0.0746 - accuracy: 0.9764


The results of the training can be now evaluated on a given test set.

In [13]:
model.evaluate(x_test,  y_test, verbose=2)

10000/10000 - 1s - loss: 0.0750 - accuracy: 0.9781


[0.07497456320775672, 0.9781]

We have trained an image classifier on the MNIST dataset with an accuracy of around 98%.

Now we pri